WordPress/Learn
# Workshops: Add an application form and workflow

Adds a custom form for applying to present a workshop. The form fields are based on the [existing form](https://learn.wordpress.org/workshop-presenter-application/) which was built with Jetpack's Contact Form block. This form, however, doesn't require the applicant to enter info like name and email, but instead requires that they log into their wp.org account first, where that data is already available.

The form is output by a `wporg-learn/workshop-application-form` custom block. Styles for the form front end are added to the Learn theme stylesheet. Field validation errors are shown inline with the fields they pertain to.

On the back end, a successful form submission will create a new workshop post with the "Needs Vetting" post status, which is set up via the Edit Flow plugin. Some of the application fields are used to populate the workshop post, while others are just information for the application reviewer. All of the fields from the original application are output in a metabox at the bottom of the Edit Workshop screen.

This is the third recent project I've done that called for validation of some kind of structured data, so I decided to build a tool that can flexibly handle a lot of different use cases, so I can stop reinventing the wheel every time. The result is the `\WordPressdotorg\Validator` class, which lives in wp.org's mu-plugins. Provide the validator instance with a schema for the structured data, run some data through it, and it will output detailed information about anything invalid about the data.

Refs #23 

## Screenshots

![form1](https://user-images.githubusercontent.com/916023/95506974-915bee00-0965-11eb-8fde-be414fcd0f83.jpg)

![form2](https://user-images.githubusercontent.com/916023/95506989-9751cf00-0965-11eb-88fe-929766e140df.jpg)

## To test

1. Make sure Edit Flow is installed and activated. You can rebuild your local environment with `yarn run create` which should take care of it.
1. Configure Edit Flow. The Custom Statuses module needs to be enabled and applied to the Workshops post type. A "Needs Vetting" status needs to be created.
1. Create a page for the application form to live. Add the Workshop Application Form block to the page and publish.
1. View the new page. While logged in, the application form should be visible. While logged out, you should be prompted to log in.
1. Try to break the form while filling it out. Of particular note are the "Other" open text inputs on the audience and experience level questions.
1. After getting a successful form submission, view the Workshops list table and find the newly created post. The title should be what you put in the "Workshop Title" field. The status should be "Needs Vetting".
1. The new workshop post should have content for the description, learning objectives, and comprehension questions. The language should be added to the Workshop Details metabox. The username of the form submitter should be in the Presenters metabox. All of the form input values should be in an "Original Application" metabox at the bottom of the screen.
